### PR TITLE
ADIOS2 Append Mode

### PR DESCRIPTION
**Pull request by franzpoeschel**
Our `READ_WRITE` mode is not quite adequate for use in ADIOS2, since ADIOS2 only allows either reading or writing, but our `READ_WRITE` mode's workflow needs that both are working (#996). However, Append mode exposes an interface equivalent to that of the Write mode, but steps already on disk are not deleted, new ones are just appended. Hence, add a new `Access::APPEND` that unlike `READ_WRITE` does not allow any reading, but does not overwrite anything either.

- [x] Simple proof-of-concept implementation
- [x] Test all possible edge cases:
    * Since there is no reading involved, it's on the user to use the exact same configuration of openPMD in append mode as had been used for the appended-to file.
    * This is now testing new and old ADIOS schema, group- file- and variablebased encoding
    * Overwriting old iterations is still a work-in-progress, non-linear read patterns and truncation of files upon opening are to come in a later PR
- [x] Keep possible a workflow where in file-based iteration encoding, already existing iterations can be read and new ones can be written
  The `READ_WRITE` mode currently does this. I would suggest to use new Append mode to support this more properly. Will need some tweaks in the frontend though where APPEND mode is currently treated as equivalent to WRITE mode.
  Idea: In filebased iteration encoding, coerce the backendAccess to CREATE. Any written iteration will then be overwritten in doubt.
- [x] Bring this to the other backends. This should become the new default `READ_WRITE` mode light, unless you really need to read data.
- [ ] Discussion point: Since reading does not work *at all* in this mode, I don't see any way to guarantee that an update to a dataset is written with the same overall settings (schema, usage of steps, ..). Users will need to make sure to create consistent datasets...
- [ ] Discussion point: Checkpoint-restarting. Probably one of the main uses of this. Dump every 100 time steps, checkpoint every 1000, crash at 1500, restart from 1000. Consequence: Steps {1000,1100,…,1500} are written twice. See #949 for an exploration on how to deal with this, but it's not really ideal.
    Update: See https://github.com/ornladios/ADIOS2/issues/2775
    Will still not be trivial to implement, since a preceding read access might be necessary to figure out how many steps should be dropped. Would also require to specify from which iteration to restart writing.
- [x] Documentation page on access modi
- [x] File-based appending in ADIOS2: Throw error if trying to append to already written iteration?
    Update: Given the proposed workflow with https://github.com/ornladios/ADIOS2/issues/2775, it would be most consistent to just overwrite iterations. Maybe use `Access::CREATE` again for the backendAccess.
- [x] ADIOS1 UPDATE: OperationUnsupportedInBackend
- [x] Merge #1080 first
- [x] Document altered meaning of `createFile` task

Also, this introduces a new `Error.hpp`, so I don't need to again throw `std::runtime_errors` around.

### CODE REVIEW

**Comment by ax3l** [comment]
> Checkpoint-restarting. Probably one of the main uses of this. Dump every 100 time steps, checkpoint every 1000, crash at 1500, restart from 1000. Consequence: Steps {1000,1100,…,1500} are written twice. See #949 for an exploration on how to deal with this, but it's not really ideal.

I think useful would be a mode that overwrites a step exactly when it is re-opened again, but leaves older (and even newer) steps as they are. That's what people most expect and know from file-based output: files stay around until you overwrite one by one.

Since output from checkpoint-restart should give again the same result as the prior run, this is also ok (no "invalid" older files, so to say).

Alternatively, one could at a certain point (e.g. once during re-open/append) delete all steps _after_ the restart step.

**Comment by franzpoeschel** [comment]
> I think useful would be a mode that overwrites a step exactly when it is re-opened again, but leaves older (and even newer) steps as they are. That's what people most expect and know from file-based output: files stay around until you overwrite one by one.

That is possible to implement with file-based iteration encoding and APPEND mode, but it won't be possible with either group- or variable-based encoding. Adding steps to an existing file in ADIOS2 allows no reading or modifying of old steps, you're just appending to a BP file without caring what had previously been written to it. 

The problem is not file-based encoding, but variable-/step-based encodings with steps enabled *for output data* after loading a checkpoint.

File: include/openPMD/IO/AbstractIOHandler.hpp [hunk 1]
```diff
@@ -97,6 +97,25 @@ enum class FlushLevel : unsigned char
...
+         * In file-based iteration encoding, the APPEND mode is handled entirely
+         * by the frontend, the backend should just treat it as CREATE mode
+         * @todo maybe replace the encoding parameter of the OPEN_FILE task
+         *     with this?
+         */
+        if( encoding == IterationEncoding::fileBased &&
+            m_backendAccess == Access::APPEND )
+        {
+            // do we really want to have those as const members..?
+            *const_cast< Access * >( &m_backendAccess ) = Access::CREATE;
```
Comment (franzpoeschel) [inline include/openPMD/IO/AbstractIOHandler.hpp]: I have now added some handling in the backend so that this isn't necessary any more. Could remove it.

**Comment by ax3l** [comment]
> The problem is not file-based encoding, but variable-/step-based encodings with steps enabled for output data after loading a checkpoint.

Uff, right. I put this on the ADIOS2 agenda. I think we could try to gracefully warn and skip such outputs in user-code if they exist, but definitely have to think about a workflow.

**Comment by franzpoeschel** [comment]
~~I quickly deleted the member and looked where compilation failed. It seems that there currently is no other bug of the sort.~~ wrong PR

**Comment by ax3l** [comment]
Discussion update:
- RFE with ADIOS2: Append mode, with parameter to truncate after the Nth (ADIOS) step: https://github.com/ornladios/ADIOS2/issues/2775
- high-level solution: using the openPMD “step” variable to list the same step again and our logic picks up the “latest” one in reads (will work, but accumulates data if steps are written multiple times)

**Comment by franzpoeschel** [comment]
I think this is ready for review @ax3l 
Note that there is further info found in the added documentation, otherwise I have tried to split this into a sane number of commits.

File: include/openPMD/IO/AbstractIOHandler.hpp (hunk 1 above)
Comment (franzpoeschel) [reply include/openPMD/IO/AbstractIOHandler.hpp]: Update: Eh, let' keep it. Clean overwriting is better and in file-based encoding it's easy to achieve.

**Comment by franzpoeschel** [comment]
I've done further cleanup now. If it runs green, ready for review @ax3l

File: docs/source/usage/workflow.rst [hunk 2]
```diff
@@ -1,5 +1,38 @@
 .. _workflow:
 
+Access modes
+============
+
+The openPMD-api distinguishes between a number of different access modes:
+
+* **Create mode**: Used for creating a new Series from scratch.
+  Any file possibly existing in the specified location will be overwritten.
```
Comment (ax3l) [inline docs/source/usage/workflow.rst:9]: For file-based I/O, I think there might be a potentially confusing corner case:
if one writes the same series again, but with different interval, then some files might be overwritten and others untouched.

I am thinking about this from time to time and see pro and cons for, e.g., actively cleaning/moving out all potential files (even for iterations that are not created again) when _opening_ in truncating create mode.

More of a thing we could discuss next time.

File: docs/source/usage/workflow.rst [hunk 3]
```diff
@@ -1,5 +1,38 @@
...
+* **Read-only mode**: Used for reading from an existing Series.
+  No modifications will be made.
+* **Read/Write mode**: Creates a new Series if not existing, otherwise opens an existing Series for reading and writing.
+  New datasets and iterations will be inserted as needed.
+  Not fully supported by all backends:
+  * ADIOS1: Automatically coerced to *Create* mode if the file does not exist yet and to *Read-only* mode if it exists.
```
Comment (ax3l) [inline docs/source/usage/workflow.rst:16]: rst lists are weird: newline needed before starting them:
```suggestion

  * ADIOS1: Automatically coerced to *Create* mode if the file does not exist yet and to *Read-only* mode if it exists.
```

File: docs/source/usage/workflow.rst [hunk 4]
```diff
@@ -1,5 +1,38 @@
...
+  * ADIOS2: Automatically coerced to *Create* mode if the file does not exist yet and to *Read-only* mode if it exists.
+    Since this happens on a per-file level, this mode allows to read from existing iterations and write to new iterations at the same time in file-based iteration encoding.
+* **Append mode**: Restricted mode for appending new iterations to an existing Series that is supported by all backends at least in file-based iteration encoding, and by all but ADIOS1 in other encodings.
+  The API is equivalent to that of the *Create* mode, meaning that no reading is supported whatsoever.
+  If the Series does not exist yet, this behaves equivalently to the *Create* mode.
+  Existing iterations will not be deleted, newly-written iterations will be inserted.
+
+  **Warning:** If writing an iteration that already exists, the behavior is implementation-defined and depends on the chosen backend and iteration encoding:
```
Comment (ax3l) [inline docs/source/usage/workflow.rst]:
```suggestion
  **Warning:** When writing an iteration that already exists, the behavior is implementation-defined and depends on the chosen backend and iteration encoding:
```

File: docs/source/usage/workflow.rst [hunk 5]
```diff
@@ -1,5 +1,38 @@
...
+* **Create mode**: Used for creating a new Series from scratch.
+  Any file possibly existing in the specified location will be overwritten.
+* **Read-only mode**: Used for reading from an existing Series.
+  No modifications will be made.
+* **Read/Write mode**: Creates a new Series if not existing, otherwise opens an existing Series for reading and writing.
+  New datasets and iterations will be inserted as needed.
+  Not fully supported by all backends:
+  * ADIOS1: Automatically coerced to *Create* mode if the file does not exist yet and to *Read-only* mode if it exists.
+  * ADIOS2: Automatically coerced to *Create* mode if the file does not exist yet and to *Read-only* mode if it exists.
+    Since this happens on a per-file level, this mode allows to read from existing iterations and write to new iterations at the same time in file-based iteration encoding.
```
Comment (ax3l) [inline docs/source/usage/workflow.rst]:
```suggestion
  Since this happens on a per-file level, this mode allows to read from existing iterations and write to new iterations at the same time in file-based iteration encoding.
```

File: docs/source/usage/workflow.rst [hunk 6]
```diff
@@ -1,5 +1,38 @@
...
+  Existing iterations will not be deleted, newly-written iterations will be inserted.
+
+  **Warning:** If writing an iteration that already exists, the behavior is implementation-defined and depends on the chosen backend and iteration encoding:
+
+  * The new iteration might fully replace the old one.
+  * The new iteration might be merged into the old one.
+  * (To be removed in a future update) The old and new iteration might coexist in the resulting dataset.
+
+  We suggest to fully define iterations when using Append mode (i.e. as if using Create mode) to avoid implementation-specific behavior.
+  Appending to an openPMD Series is only supported on a per-iteration level.
```
Comment (ax3l) [inline docs/source/usage/workflow.rst:31]: Maybe a bit too strict here to understand what is meant or might be repetitive with what is said above?
Because one could add here: `Extending the shape of existing data sets is possible, see the examples.`

File: include/openPMD/Error.hpp [hunk 7]
```diff
@@ -62,5 +62,17 @@ namespace error
     public:
         WrongAPIUsage( std::string what );
     };
+
+    /**
+     * @brief Internal errors that should not happen. Please report.
+     *
+     * Example: A nullpointer is observed somewhere.
+     */
+
```
Comment (ax3l) [inline include/openPMD/Error.hpp]:
```suggestion
```

File: include/openPMD/IO/AbstractIOHandler.hpp [hunk 8]
```diff
@@ -131,6 +148,7 @@ class AbstractIOHandler
     virtual std::string backendName() const = 0;
 
     std::string const directory;
+    // why do these need to be separate?
```
Comment (ax3l) [inline include/openPMD/IO/AbstractIOHandler.hpp:173]: Do we want to answer this question? :)

I am not sure we documented this well when this was introduced (also visible in the missing doxygen strings for those two members).

File: src/IO/JSON/JSONIOHandlerImpl.cpp [hunk 9]
```diff
@@ -1077,6 +1078,14 @@ namespace openPMD
         {
             case Access::CREATE:
             case Access::READ_WRITE:
+            case Access::APPEND:
+                /*
+                 * Always truncate when writing, we alway write entire JSON
+                 * datasets, never partial ones.
```
Comment (ax3l) [inline src/IO/JSON/JSONIOHandlerImpl.cpp]: Writing with `storeChunk` can still be _partial_, even in serial writes. Thus the text here is a bit unclear in the argument, I think.

(I am ok with the logic to truncate.)

File: test/SerialIOTest.cpp [hunk 10]
```diff
@@ -1629,8 +1629,24 @@ void fileBased_write_test(const std::string & backend)
         REQUIRE(o.iterations[5].time< double >() == 5.0);
     }
 
-    // extend existing series with new step and auto-detection of iteration padding
+    if( backend == "bp" )
     {
+        // Append + filebased iteration encoding works for all backends
```
Comment (ax3l) [inline test/SerialIOTest.cpp]: The `if "bp"` above is confusing (outdated?) if we want to run this for all backends now :)

File: test/SerialIOTest.cpp [hunk 11]
```diff
@@ -1673,6 +1689,7 @@ void fileBased_write_test(const std::string & backend)
     }
 
     // write with auto-detection and in-consistent padding
+    if( backend != "bp" )
```
Comment (ax3l) [inline test/SerialIOTest.cpp]: Double-checking: still skip for `bp`?

File: include/openPMD/IO/AbstractIOHandler.hpp (hunk 8 above)
Comment (guj) [reply include/openPMD/IO/AbstractIOHandler.hpp:173]: I ran into that a few times. I believe it would be consistent to treat all files in the file mode as they were one file. When its time to overwrite, remove all existing files related to the  series.

File: src/IO/ADIOS/ADIOS2IOHandler.cpp [hunk 12]
```diff
@@ -1063,21 +1061,16 @@ ADIOS2IOHandlerImpl::adios2AccessMode( std::string const & fullPath )
         if( auxiliary::directory_exists( fullPath ) ||
             auxiliary::file_exists( fullPath ) )
         {
-            std::cerr << "ADIOS2 does currently not yet implement ReadWrite "
-                         "(Append) mode. "
-                      << "Replacing with Read mode." << std::endl;
             return adios2::Mode::Read;
```
Comment (guj) [inline src/IO/ADIOS/ADIOS2IOHandler.cpp:1074]: Now that the APPEND mode  is added.  Should READ_WRITE mode be rejected for ADIOS?  Might be semantically cleaner.

File: docs/source/usage/workflow.rst (hunk 6 above)
Comment (franzpoeschel) [reply docs/source/usage/workflow.rst:31]: In that case, let's maybe restrict the wording to make clear that this refers to the Append mode only:

> The Append mode is intended for adding new iterations to a Series, not for the modification of existing iterations.

File: include/openPMD/IO/AbstractIOHandler.hpp (hunk 8 above)
Comment (franzpoeschel) [reply include/openPMD/IO/AbstractIOHandler.hpp:173]: > Do we want to answer this question? :)

I'd suggest to do it in a separate PR, need to figure out what the answer is at first though
I wasn't always quite sure which one of these to use when writing this PR

> I ran into that a few times. I believe it would be consistent to treat all files in the file mode as they were one file. When its time to overwrite, remove all existing files related to the series.

We could think about this for CREATE mode, since its semantics are "overwrite anything that exists". The purpose READ_WRITE and APPEND modes is explicitly *not* to delete existing data. For APPEND mode, the most sensible solution seems to be truncation (i.e. "delete anything past iteration 500") which would be a follow-up PR.

File: src/IO/ADIOS/ADIOS2IOHandler.cpp (hunk 12 above)
Comment (franzpoeschel) [reply src/IO/ADIOS/ADIOS2IOHandler.cpp:1074]: READ_WRITE is still useful for file-based iteration encoding since it allows you to read old iterations while at the same time to write new ones. This feature is not found in any other iteration encodings. For the other iteration encodings, there is no use in having READ_WRITE mode for ADIOS2, yeah.
We should however not deactivate it immediately, but allow users to transition to APPEND mode since it's a breaking change.

File: src/IO/JSON/JSONIOHandlerImpl.cpp (hunk 9 above)
Comment (franzpoeschel) [reply src/IO/JSON/JSONIOHandlerImpl.cpp]: What I mean in here is that this backend always re-writes the entire JSON file and does not just update parts of it. This makes it ok to just truncate the file. So maybe:
```suggestion
                 * Always truncate when writing, we alway write entire JSON
                 * files, never partial ones.
```

File: test/SerialIOTest.cpp (hunk 10 above)
Comment (franzpoeschel) [reply test/SerialIOTest.cpp]: Outdated, yeah
While working on the PR, read_write mode was for a while completely inaccessible for ADIOS2 because I had planned to replace it with Append mode entirely
Then I noticed that there are still reasons to use read_write with ADIOS2 in file-based iteration encoding, so those specialized test paths can now be deleted again

**Comment by ax3l** [comment]
@franzpoeschel Just for after the holidays: This one needs a rebase please :)

The first commit (`error::Internal`) is probably already merged via #1037, and a few of the other commits have tiny merge conflicts, I think.

**Comment by franzpoeschel** [comment]
> @franzpoeschel Just for after the holidays: This one needs a rebase please :)
> 
> The first commit (`error::Internal`) is probably already merged via #1037, and a few of the other commits have tiny merge conflicts, I think.

I've rebased this now. The introduced error types are different ones.

File: test/SerialIOTest.cpp (hunk 11 above)
Comment (franzpoeschel) [reply test/SerialIOTest.cpp]: I've removed the skip now

File: docs/source/usage/workflow.rst (hunk 2 above)
Comment (franzpoeschel) [reply docs/source/usage/workflow.rst:9]: Fair point. I guess the most sensible solution would be within the bounds of a potential truncate option.

**Comment by ax3l** [comment]
@franzpoeschel please rebase due to clang-format changes

You might find this helpful: https://github.com/pre-commit/pre-commit/issues/1696#issuecomment-974621448 :)

**Comment by franzpoeschel** [comment]
> @franzpoeschel please rebase due to clang-format changes

Done :)
> 
> You might find this helpful: [pre-commit/pre-commit#1696 (comment)](https://github.com/pre-commit/pre-commit/issues/1696#issuecomment-974621448) :)

I've used [this script](https://github.com/franzpoeschel/reformat-rebase) which entirely avoids pre-commit getting active at all

**Comment by franzpoeschel** [comment]
clang-6 is being a bit weird with checking for exceptions:
```
 /home/runner/work/openPMD-api/openPMD-api/test/SerialIOTest.cpp:5573: FAILED:
  REQUIRE_THROWS_AS( write.flush(), error::OperationUnsupportedInBackend )
due to unexpected exception with message:
  Operation unsupported in ADIOS1: Appending to existing file on disk (use
  Access::CREATE to overwrite)
```

And the Windows / MSVC errors I don't understand at all tbf
But I'll try to have a look

UPDATE:

The MSVC errors seem to stem from pybind: https://github.com/microsoft/onnxruntime/issues/9735#issuecomment-969577865
>everybody using pybind11 with VS2022 and STL should hit the issue.

File: share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h [hunk 13]
```diff
@@ -10,6 +10,24 @@
 
 #pragma once
 
+/*
```
Comment (franzpoeschel) [inline share/openPMD/thirdParty/pybind11/include/pybind11/pybind11.h]: This one seems to fix the CI issues that I saw. Might want to put it in a separate PR and merge it with the MSVC block below.

**Comment by franzpoeschel** [comment]
Many CI runs currently show this behavior:
```
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
read missing: successfully caught! json
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
read missing: successfully caught! h5
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 0:
  #000: ../../../src/H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: ../../../src/H5Fint.c line 1400 in H5F__open(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: ../../../src/H5Fint.c line 1546 in H5F_open(): unable to open file: time = Wed Apr  6 14:19:34 2022
, name = '../samples/git-sample/data00000100.h5', tent_flags = 1
    major: File accessibilty
    minor: Unable to open file
  #003: ../../../src/H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #004: ../../../src/H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = '../samples/git-sample/data00000100.h5', errno = 13, error message = 'Permission denied', flags = 1, o_flags = 2
    major: File accessibilty
    minor: Unable to open file
openPMD series: empty_alternate_fbpic_%T
openPMD standard: 1.0.0
openPMD extensions: 1

number of iterations: 1

number of meshes: 4

number of particle species: 1
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
[AbstractIOHandlerImpl] IO Task CLOSE_FILE failed with exception. Removing task from IO queue and passing on the exception.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SerialIOTests is a Catch v2.13.4 host application.
Run with -? for options

-------------------------------------------------------------------------------
git_hdf5_sample_fileBased_read_test
-------------------------------------------------------------------------------
/home/runner/work/openPMD-api/openPMD-api/test/SerialIOTest.cpp:2975
...............................................................................

/home/runner/work/openPMD-api/openPMD-api/test/SerialIOTest.cpp:2975: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  [HDF5] Trying to close a file that is not present in the backend
```
No idea what's going on yet, will have a look

**Comment by franzpoeschel** [comment]
Ok, I get nearly exactly the same error message locally when I remove my reading rights via `chmod a-r`:
```
Filters: git_hdf5_sample_fileBased_read_test
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 0:
  #000: ../../../src/H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: ../../../src/H5Fint.c line 1400 in H5F__open(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: ../../../src/H5Fint.c line 1546 in H5F_open(): unable to open file: time = Thu Apr  7 10:52:21 2022
, name = '../samples/git-sample/data00000100.h5', tent_flags = 0
    major: File accessibilty
    minor: Unable to open file
  #003: ../../../src/H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #004: ../../../src/H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = '../samples/git-sample/data00000100.h5', errno = 13, error message = 'Permission denied', flags = 0, o_flags = 0
    major: File accessibilty
    minor: Unable to open file
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
[AbstractIOHandlerImpl] IO Task CLOSE_FILE failed with exception. Removing task from IO queue and passing on the exception.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SerialIOTests is a Catch v2.13.4 host application.
Run with -? for options

-------------------------------------------------------------------------------
git_hdf5_sample_fileBased_read_test
-------------------------------------------------------------------------------
/home/franzpoeschel/git-repos/openPMD-api/test/SerialIOTest.cpp:2975
...............................................................................

/home/franzpoeschel/git-repos/openPMD-api/test/SerialIOTest.cpp:2975: FAILED:
due to unexpected exception with message:
  [HDF5] Trying to close a file that is not present in the backend
```

Seems like a further CI issue to me..

**Comment by ax3l** [comment]
I think the problem here is usually if we try to open a file in "write" mode, although it should be read-only.
We set HDF5 files to "read-only" with `chmod` to catch that bug: #394 / #460 - we want to make sure that we do not accidentally trigger "write" mode if we only read (`+r-w` permissions _must_ be sufficient).

**Comment by franzpoeschel** [comment]
> I think the problem here is usually if we try to open a file in "write" mode, although it should be read-only. We set HDF5 files to "read-only" with `chmod` to catch that bug: #394 / #460 - we want to make sure that we do not accidentally trigger "write" mode if we only read (`+r-w` permissions _must_ be sufficient).

Ah, so there's a potential bug to hunt down. Will check it.

**Comment by franzpoeschel** [comment]
Ok, by taking my writing permissions I can reproduce the same issue offline:

```
Filters: git_hdf5_sample_fileBased_read_test
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 0:
  #000: ../../../src/H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: ../../../src/H5Fint.c line 1400 in H5F__open(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: ../../../src/H5Fint.c line 1546 in H5F_open(): unable to open file: time = Fri Apr 29 14:11:16 2022
, name = '../samples/git-sample/data00000100.h5', tent_flags = 1
    major: File accessibilty
    minor: Unable to open file
  #003: ../../../src/H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #004: ../../../src/H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = '../samples/git-sample/data00000100.h5', errno = 13, error message = 'Permission denied', flags = 1, o_flags = 2
    major: File accessibilty
    minor: Unable to open file
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
[AbstractIOHandlerImpl] IO Task CLOSE_FILE failed with exception. Removing task from IO queue and passing on the exception.
```

However, the failing line is this one:
```cpp
            Series o =
                Series("../samples/git-sample/data%T.h5", Access::READ_WRITE);
```
So, read-write access mode is entirely justified?

**Comment by franzpoeschel** [comment]
The `dev` branch gives me the same error, but lets the test pass:

```
Filters: git_hdf5_sample_fileBased_read_test
HDF5-DIAG: Error detected in HDF5 (1.10.4) thread 0:
  #000: ../../../src/H5F.c line 509 in H5Fopen(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #001: ../../../src/H5Fint.c line 1400 in H5F__open(): unable to open file
    major: File accessibilty
    minor: Unable to open file
  #002: ../../../src/H5Fint.c line 1546 in H5F_open(): unable to open file: time = Fri Apr 29 14:23:12 2022
, name = '../samples/git-sample/data00000100.h5', tent_flags = 1
    major: File accessibilty
    minor: Unable to open file
  #003: ../../../src/H5FD.c line 734 in H5FD_open(): open failed
    major: Virtual File Layer
    minor: Unable to initialize object
  #004: ../../../src/H5FDsec2.c line 346 in H5FD_sec2_open(): unable to open file: name = '../samples/git-sample/data00000100.h5', errno = 13, error message = 'Permission denied', flags = 1, o_flags = 2
    major: File accessibilty
    minor: Unable to open file
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
git sample not accessible. ([HDF5] Failed to open HDF5 file ../samples/git-sample/data00000100.h5)
===============================================================================
All tests passed (13 assertions in 1 test case)
```

**Comment by franzpoeschel** [comment]
Ah, ok, something changed in the error handling and the error is replaced with a new one:
```
[AbstractIOHandlerImpl] IO Task OPEN_FILE failed with exception. Removing task from IO queue and passing on the exception.
[AbstractIOHandlerImpl] IO Task CLOSE_FILE failed with exception. Removing task from IO queue and passing on the exception.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
SerialIOTests is a Catch v2.13.4 host application.
Run with -? for options

-------------------------------------------------------------------------------
git_hdf5_sample_fileBased_read_test
-------------------------------------------------------------------------------
/home/franzpoeschel/git-repos/openPMD-api/test/SerialIOTest.cpp:2989
...............................................................................

/home/franzpoeschel/git-repos/openPMD-api/test/SerialIOTest.cpp:2989: FAILED:
  {Unknown expression after the reported line}
due to unexpected exception with message:
  [HDF5] Trying to close a file that is not present in the backend
```

I'll check if I can fix that some way.

**Comment by franzpoeschel** [comment]
Ready for review again :) @ax3l

File: src/IO/HDF5/HDF5IOHandler.cpp [hunk 14]
```diff
@@ -409,6 +436,46 @@ void HDF5IOHandlerImpl::createDataset(
             "[HDF5] Internal error: Failed to open HDF5 group during dataset "
             "creation");
 
+        if (m_handler->m_backendAccess == Access::APPEND)
+        {
+            // The dataset might already exist in the file from a previous run
+            // We delete it, otherwise we could not create it again with
+            // possibly different parameters.
+            // This is inefficient, as only the link to the dataset will be
+            // removed, but not the actual dataset
```
Comment (franzpoeschel) [inline src/IO/HDF5/HDF5IOHandler.cpp]: Is it inefficient? I don't really know about HDF5's garbage collection mechanics. Do they need to be triggered manually? Do they apply in this situation?

**Review by ax3l — APPROVED** [review]
This looks great, thank you :tada:

**Comment by ax3l** [comment]
Should we clarify the two discussion points first in the PR description?
first one: I guess yes.
second one: moved to https://github.com/ornladios/ADIOS2/issues/2775

File: src/IO/HDF5/HDF5IOHandler.cpp (hunk 14 above)
Comment (ax3l) [reply src/IO/HDF5/HDF5IOHandler.cpp]: I think they are unlinked and not removed unless one does a HDF5 repack. I think that's also the right thing to do.

maybe @jeanbez knows is the below code looks solid to remove a HDF5 object? :)

File: src/IO/HDF5/HDF5IOHandler.cpp [hunk 15]
```diff
@@ -409,6 +436,46 @@ void HDF5IOHandlerImpl::createDataset(
...
+                    node_id, i, charbuffer.data(), name_length + 1);
+                if (std::strncmp(
+                        name.c_str(), charbuffer.data(), name_length + 1) != 0)
+                {
+                    continue;
+                }
+                status = H5Ldelete(node_id, name.c_str(), H5P_DEFAULT);
+                VERIFY(
+                    status == 0,
+                    "[HDF5] Internal error: Failed to delete old dataset from "
```
Comment (ax3l) [inline src/IO/HDF5/HDF5IOHandler.cpp]: If you include `name` in the error message, the error gets more useful for users.

File: src/Error.cpp [hunk 16]
```diff
@@ -45,5 +45,9 @@ namespace error
               concatVector(errorLocation_in) + "': " + std::move(what))
         , errorLocation(std::move(errorLocation_in))
     {}
+
+    Internal::Internal(std::string const &what)
+        : Error("Internal error: " + what + "\nThis is a bug. Please report.")
```
Comment (ax3l) [inline src/Error.cpp]: add the URL where to report here?
`https://github.com/openPMD/openPMD-api/issues`

File: src/IO/HDF5/HDF5IOHandler.cpp [hunk 17]
```diff
@@ -237,13 +237,40 @@ void HDF5IOHandlerImpl::createFile(
...
+            }
+            break;
+        case Access::READ_WRITE:
             flags = H5F_ACC_EXCL;
-        hid_t id =
-            H5Fcreate(name.c_str(), flags, H5P_DEFAULT, m_fileAccessProperty);
+            break;
+        case Access::READ_ONLY:
+            // condition has been checked above
+            throw std::runtime_error("Control flow error");
```
Comment (ax3l) [inline src/IO/HDF5/HDF5IOHandler.cpp]:
```suggestion
            throw std::runtime_error("[HDF5] Control flow error in createFile backend access mode.");
```

File: src/IO/HDF5/HDF5IOHandler.cpp (hunk 14 above)
Comment (franzpoeschel) [reply src/IO/HDF5/HDF5IOHandler.cpp]: > I think that's also the right thing to do.

So you mean we should repack here?

File: src/Error.cpp [hunk 18]
```diff
@@ -45,5 +45,12 @@ namespace error
               concatVector(errorLocation_in) + "': " + std::move(what))
         , errorLocation(std::move(errorLocation_in))
     {}
+
+    Internal::Internal(std::string const &what)
+        : Error(
+              "Internal error: " + what +
+              "\nThis is a bug. Please report at ' "
+              "https://github.com/openPMD/openPMD-api/issues'.")
```
Comment (ax3l) [inline src/Error.cpp:53]: Hard to copy, did read `' url'.`
```suggestion
              "\nThis is a bug. Please report at "
              "https://github.com/openPMD/openPMD-api/issues")
```

File: src/IO/HDF5/HDF5IOHandler.cpp (hunk 14 above)
Comment (ax3l) [reply src/IO/HDF5/HDF5IOHandler.cpp]: No, repack is an expensive operation one does in post. It copies everything.
Just calling delete here is the right thing to do. It might actually free disk space, but in doubt it is ok if it does not (and only hides the entry).

**Review by ax3l — APPROVED** [review]

**Comment by ax3l** [comment]
@franzpoeschel please needs a rebase. Collided with #1264

File: src/IO/HDF5/HDF5IOHandler.cpp (hunk 14 above)
Comment (jeanbez) [reply src/IO/HDF5/HDF5IOHandler.cpp]: @ax3l I was not sure, so I checked with Tang/Suren and they mentioned that when removing an object with H5Ldelete, the HDF5 library should be able to detect and recycle the file space when no other reference to the deleted object exists. Regarding the code below it's more efficient to do H5Lexists to test whether a group/dset/attr exists, than iterating over existing items in a group like the one in this code snippet.

Comment (ax3l) [reply src/IO/HDF5/HDF5IOHandler.cpp]: Thanks a lot for the details, @jeanbez.

Can we also simply unlink a group and HDF5 garbage collects all underlying inodes of that group? (Similar for data sets?) The challenge we have here is that a user might have added custom attributes that we do not know of.

Comment (franzpoeschel) [reply src/IO/HDF5/HDF5IOHandler.cpp]: The latest commit [7afb99d](https://github.com/openPMD/openPMD-api/pull/1007/commits/7afb99dbb71bde944a42969db99122732524222f) now uses H5Lexists, thanks for the tip!

Comment (jeanbez) [reply src/IO/HDF5/HDF5IOHandler.cpp]: @ax3l Yes, according to the HDF5 team the refcount will be decremented on the group when it's unlinked from its parent, and the group will be deleted if its refcount drops to zero.  If a group is deleted, it will unlink all of its children, possibly causing them to be deleted as well.

Comment (jeanbez) [reply src/IO/HDF5/HDF5IOHandler.cpp]: Update: empty space in the file can only be reused until the file is closed, unless persistent free space tracking is on (it is off by default)
